### PR TITLE
Implement balt test runner

### PR DESCRIPTION
**Pull request by manuranga**
Add balt tests for subset-01

Part of #358

### CODE REVIEW

**Comment by jclark** [comment]
I suggest we split this up into:

1. changes to the driver to support .balt input files
2. converting tests from current format into .balt

We can merge (1) before we have done (2).

**Comment by manuranga** [comment]
I have change the ll file to have both an incremental number and line number. Makefile's incremental/parallel testing is mostly working, I need few adjustments.

```
out
├── compile.stamp
├── 01-boolean.stamp
├── ...
├── expect
│   ├── 01-boolean
│   │   ├── V01-boolean-0000-0001.txt
│   │   ...
│   │   └── V01-boolean-0010-0179.txt
├── ll
│   ├── 01-boolean
│   │   ├── V01-boolean-0000-0001.txt
```

Should I organize `out` dir to be  kind-major (`out/expect/01-boolean/V01-boolean-0000-0001.txt`) or to be balt-name-major (`out/01-boolean/expect/V01-boolean-0000-0001.txt`) ?

**Comment by jclark** [comment]
Is there any need to repeat eg 01-boolean in the filename as well as the directory name?

**Comment by jclark** [comment]
Should be out/01-boolean

**Comment by manuranga** [comment]
> Is there any need to repeat eg 01-boolean in the filename as well as the directory name?

No need, it's there due to a historical reason, will remove.

>Should be out/01-boolean

Ack

**Comment by jclark** [comment]
Another thing to keep in mind: we will soon want to support multiple files per module; when we do so, the meaning of multiple `.bal` files on the command-line should be to treat them as all being part of the same module.  It would be good if we can test this with `.balt` files.

Also we are soon going to want to add support for multiple modules, which we will mean we need test cases with multiple modules. We will need to be able to test using `.balt` files.

No need to do either of these things right now, but we should design so we can easily evolve to support them.

**Comment by jclark** [comment]
I am going to attempt multiple source files #365 now.

**Comment by manuranga** [comment]
Getting `out/01-boolean/expect/V0000L0001.txt` patten needed some fiddling around because you can't write
`out/%/expect/%.txt` as a rule. Adding `.SECONDEXPANSION` did the trick.

I will test things a bit more and send it soon.

**Comment by jclark** [comment]
We shouldn't need a rule `out/%/expect/%.txt`. We should be doing a recursive make in `out/01-boolean`.

**Comment by manuranga** [comment]
I updated the PR, since it's working and is incremental/parallel, and ready to be reviewed.
I still need to switch to recursive make approach for testing sub dirs.

**Comment by jclark** [comment]
I reworked `main.bal` a lot as part of #365. It had become a complete mess.

File: compiler/balt.bal [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        hash = fnv1a32(l.toBytes(), hash);
+    }
+    return hash;
+}
+
+function fnv1a32(byte[] bs, int init = FNV1_32_INIT) returns int {
+    int hash = init;
+    foreach byte b in bs {
+        hash ^= b;
+        hash *= FNV1_32_PRIME;
```
Comment (jclark) [inline compiler/balt.bal]: Isn't this going to overflow?

File: compiler/balt.bal [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            expect.push("panic: line " + (i + 1).toString() + ": " + msg);
+        }
+    }
+    return expect;
+}
+
+const FNV1_32_INIT = 0x811c9dc5;
+const FNV1_32_PRIME = 0x1000193;
+
+function hash(string[] lines) returns int {
```
Comment (jclark) [inline compiler/balt.bal]: I thought we were not hashing any more.

File: compiler/balt.bal (hunk 1 above)
Comment (manuranga) [reply compiler/balt.bal]: No, because hash is at most `0xffffffff`

File: compiler/balt.bal (hunk 2 above)
Comment (manuranga) [reply compiler/balt.bal]: Ya, will remove.

**Comment by manuranga** [comment]
> I reworked main.bal a lot as part of #365. It had become a complete mess.

Ya, I also noticed and did some refactoring, but unfortunately now it's fully conflicted with current main. Easier to get yours and rework form there.

**Comment by manuranga** [comment]
I temporary move all test bal files to files named by it's subset so we can get this pr merged.
Please use GitHub `Select commit` feature to review.

We may need a way to distinguish `FV` from `FE`, for now I am skipping everything with a `Fail-Issue:`

**Comment by jclark** [comment]
The Makefile can be done more straightforwardly using recursion as follows.

Leave the current Makefile as is for now and start off with two new Makefiles. Let's call them new.mk and sub.mk: new.mk is the one that will replace the current Makefile when we are finished; sub.mk is called by new.mk. For each balt file `NN-foo.balt` there will be an output directory `out/NN-foo/`. The job of new.mk is to run sub.mk on each of these output directories in parallel. The current directory for sub.mk will be the per-balt output directory (this is why it is less confusing to keep it separate).  Here's how you can do the recursion:

new.mk looks like this:

```
TARGETS=test all compile testll

# Construct this list with the GNU make wildcard function
TDIRS=01-boolean 01-conditional 01-function

$(TARGETS): $(TDIRS)

$(TDIRS):
	mkdir -p out/$@
	$(MAKE) -C out/$@ -f ../../sub.mk tdir=$@ $(MAKECMDGOALS)

.PHONY: $(TARGETS) $(TDIRS)
```

I used this sub.mk to test:

```
TARGETS=test all compile testll

$(TARGETS):
	@echo Starting make $@ in `pwd` with tdir=$(tdir)
	@sleep 3
	@echo Ending make $@ in `pwd` with tdir=$(tdir)

.PHONY: $(TARGETS)
```

You can derive sub.mk straightforwardly starting from current Makefile as follows:

1. remove things that are unrelated to a recursive invocation
2. changes paths to work with a current directory being the balt-specific output directory
   - some paths need ../.. added
   - some paths need to change from out/F to F
3. run the compile on a single balt file (whose name you can derive from tdir)

If you don't want to do this, I am happy to do it, but I need the changes to main.bal to support compiling balt files first.

**Comment by manuranga** [comment]
Had to add a sample test due to ballerina-platform/ballerina-lang#32575.

**Pull request merged**